external-crates/move/crates/move-compiler/src/linters/redundant_conditional.rs

```diff
@@ -0,0 +1,118 @@
+//! Detects and suggests simplification for `if c { true } else { false }` and its reverse pattern.
+//! Encourages using the condition directly (or its negation) for clearer and more concise code.
+use crate::{
+    diag,
+    diagnostics::{
+        codes::{custom, DiagnosticInfo, Severity},
+        WarningFilters,
+    },
+    expansion::ast::Value_,
+    shared::{program_info::TypingProgramInfo, CompilationEnv},
+    typing::{
+        ast::{self as T, SequenceItem_, UnannotatedExp_},
+        visitor::{TypingVisitorConstructor, TypingVisitorContext},
+    },
+};
+use move_ir_types::location::Loc;
+
+use super::{LinterDiagCategory, LINTER_DEFAULT_DIAG_CODE, LINT_WARNING_PREFIX};
+
+const REDUNDANT_CONDITIONAL_DIAG: DiagnosticInfo = custom(
+    LINT_WARNING_PREFIX,
+    Severity::Warning,
+    LinterDiagCategory::RedundantConditional as u8,
+    LINTER_DEFAULT_DIAG_CODE,
+    "",
+);
+
+pub struct RedundantConditional;
+
+pub struct Context<'a> {
+    env: &'a mut CompilationEnv,
+}
+
+impl TypingVisitorConstructor for RedundantConditional {
+    type Context<'a> = Context<'a>;
+
+    fn context<'a>(
+        env: &'a mut CompilationEnv,
+        _program_info: &'a TypingProgramInfo,
+        _program: &T::Program_,
+    ) -> Self::Context<'a> {
+        Context { env }
+    }
+}
+
+impl TypingVisitorContext for Context<'_> {
+    fn visit_exp_custom(&mut self, exp: &mut T::Exp) -> bool {
+        match &exp.exp.value {
+            UnannotatedExp_::IfElse(condition, if_block, else_block) => {
+                let condition_name: String = match condition.exp.value {
+                    UnannotatedExp_::Copy { var, .. } => {
+                        let name = var.value.name.clone();
+                        name.as_str().to_owned()
+                    }
+                    _ => "".to_owned(),
+                };
+                let extract_bool_literal_from_block = |block: &T::Exp| -> Option<bool> {
+                    if let UnannotatedExp_::Block(seq) = &block.exp.value {
```

Comment:
This doesn't have to be a block, and we should account for other cases. Immediate ones that come to mind would be just Value and Annotate (of a value)